include/lbann/utils/serialization/serialize_matrices.hpp

```diff
@@ -173,15 +240,42 @@ void load(ArchiveT& archive,
 template <typename ArchiveT, typename T,
           lbann::utils::WhenNotTextArchive<ArchiveT> = 1>
 void load(ArchiveT& archive,
-          El::Matrix<T, El::Device::GPU>& mat)
+          ::El::Matrix<T, ::El::Device::GPU>& mat)
 {
   LBANN_ASSERT(!mat.Viewing());
-  El::Matrix<T, El::Device::CPU> cpu_mat;
-  archive(cpu_mat);
-  El::Copy(cpu_mat, mat);
+  ::El::Matrix<T, ::El::Device::CPU> cpu_mat;
+  load(archive, cpu_mat);
+  ::El::Copy(cpu_mat, mat);
 }
 #endif // defined LBANN_HAS_GPU
 
+template <typename ArchiveT, typename T, ::El::Device D>
+void load(lbann::RootedInputArchiveAdaptor<ArchiveT>& ar,
+          ::El::Matrix<T,D>& mat)
+{
+  LBANN_ASSERT(!mat.Viewing());
+
+  // Restore the local matrix, then handle the Bcast
+  ar.load_on_root(mat);
+
+  // First broadcast the size information.
+  auto height = mat.Height();
+  auto width = mat.Width();
+  ::El::mpi::Broadcast(height, ar.root(), ar.grid().Comm(),
+                     ::El::SyncInfo<::El::Device::CPU>{});
+  ::El::mpi::Broadcast(width, ar.root(), ar.grid().Comm(),
+                     ::El::SyncInfo<::El::Device::CPU>{});
+  // Resize _should_ be a no-op if the size doesn't change, but I'm
+  // not actually 100% sure.
+  if (!ar.am_root())
+    mat.Resize(height, width);
+
+  // Finally the matrix data.
+  ::El::Broadcast(
+    static_cast<::El::AbstractMatrix<T>&>(mat),
+    ar.grid().Comm(), ar.root());
+}
```

Comment:
It's not deep magic, it's just C++. `static_cast` is the correct choice here because we're just working around a Hydrogen bug (normally, up-casts are implicit, but the template exists (but is undefined) for `::El::Broadcast(::El::Matrix<T,D>&, ...)`). `static_cast` is sufficiently safe and, as it's an up-cast, should be no-overhead. `reinterpret_cast` is generally not the right cast to use except, in exceptional situations, such as turning bits into data, pointers into numbers, or very low-level bit manipulations. Most authors strongly discourage its use in high-level code, and as a matter of style, I don't use it unless it's the only cast that will do what I want.